### PR TITLE
Fix: View helper should actually be a controller plugin

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] turns a view helper which should be a controller plugin into a controller plugin
* [x] implements an extension of the `ZfcUser\Controller\UserController` in order to set view variables in the view model returned from the `indexAction()`

### CODE REVIEW

**Comment by Ocramius** [comment]
@localheinz needs a rebase

**Comment by localheinz** [comment]
Rebased, @Ocramius! Thank you for the review.

File: module/User/view/zfc-user/user/index.phtml [hunk 1]
```diff
@@ -23,7 +23,6 @@
         <div class="tab-content">
             <div class="tab-pane active" id="modules">
                 <?php
-                $modules = $this->listModule(['user' => true]);
```
Comment (Ocramius) [inline module/User/view/zfc-user/user/index.phtml]: I don't actually agree with this. Why was it moved to a controller plugin? What's the rationale behind it?

Comment (localheinz) [inline module/User/view/zfc-user/user/index.phtml]: As far as I can see from what the view helper does:

* it accesses the database layer
* it makes calls to an external API
* it returns an array of data

As far as my experience goes, these are all things I would never expect from a view helper.

Instead, I would see the responsibility of a view helper in accepting some data and preparing it for presentation in the view, for example, by applying some formatting or filling the data into a view template.

Please note that, as far as I can see, I wouldn't even create a controller plugin just yet, as it's only ever used once. Instead, I'd probably leave it to a service. 

This was just the smallest refactoring possible, I believe.

**Comment by localheinz** [comment]
@Ocramius 

Also see #308.

File: module/User/view/zfc-user/user/index.phtml (hunk 1 above)
Comment (Ocramius) [inline module/User/view/zfc-user/user/index.phtml]: > these are all things I would never expect from a view helper

I disagree with this, but I also see why you'd think that. As long as these are all read operations, IMO it's fine to have them in view helpers. Yes, they sould not cause changes/events to be triggered.

 > Instead, I'd probably leave it to a service

Possibly a good solution, yes.

File: module/User/src/User/Controller/UserController.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+use ZfcUser\Controller\Plugin\ZfcUserAuthentication;
+use ZfcUser\Controller\UserController as ZfcUserController;
+
+/**
+ * @method array listModule(array $options)
+ * @method ZfcUserAuthentication zfcUserAuthentication()
+ */
+class UserController extends ZfcUserController
+{
+    public function indexAction()
```
Comment (Ocramius) [inline module/User/src/User/Controller/UserController.php]: This controller has no associated tests

**Comment by Ocramius** [comment]
Once the missing tests are introduced (integration with `@coversNothing` is fine) I can merge. I'm fine with the current approach after discussion.

**Comment by localheinz** [comment]
@Ocramius 

Added a test for the case when the authentication service has no identity.

Tried to add a test for the success case and have almost been mocking myself to death and/or trying to detach listeners, still getting failing tests because of that.

File: module/User/test/UserTest/Integration/Controller/UserControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        $serviceManager
+            ->setAllowOverride(true)
+            ->setService(
+                'zfcuser_auth_service',
+                $authenticationService
+            )
+        ;
+
+        $listModule = $this->getMockBuilder(ListModule::class)
```
Comment (Ocramius) [inline module/User/test/UserTest/Integration/Controller/UserControllerTest.php]: I think mocking `$listModule` as `function () {...}` is enough

Comment (Ocramius) [inline module/User/test/UserTest/Integration/Controller/UserControllerTest.php]: (if it makes your life easier: otherwise, current approach is fine)

Comment (localheinz) [inline module/User/test/UserTest/Integration/Controller/UserControllerTest.php]: @Ocramius 

> I think mocking `$listModule` as `function () {...}` is enough

Doesn't work, as `PluginManager` is validating and asking for an implementation of `Zend\Mvc\Controller\Plugin\PluginInterface`. However, I mocked out all of the view helpers giving me a headache, tests should pass now!

Comment (Ocramius) [inline module/User/test/UserTest/Integration/Controller/UserControllerTest.php]: Grrr @ `PluginManager`.

Yeah, mocking controllers is a mess

**Comment by Ocramius** [comment]
@localheinz let it be merged!

**Comment by localheinz** [comment]
😊

**Pull request merged**